Conditional valves (Formerly Pressure)

jellyveggie2 opened the pull request:
<!--
Pull requests must be atomic. Change one set of related things at a time.
Test your changes. PRs that were not tested will not be accepted.

You can self-label your PR. See https://ss13.moe/wiki/index.php/Guide_to_Writing_a_Pull_Request -->

## What this does
Adds some extra stuff to pressure valves, which have been renamed to conditional valves:
* You can now compare temperature instead of pressure.
* You can have them read the threshold from any of the input pipes.
* You can set any of the three pipes as the control pipe, making the side-pipe optional and allowing you to "inline" valves.

![valveui2](https://github.com/vgstation-coders/vgstation13/assets/78439377/304590f3-caff-4a2b-97ac-b7a2981265b0)

There's been some sprite changes to reflect this:

![valves](https://github.com/vgstation-coders/vgstation13/assets/78439377/e7b29476-c515-418f-9a95-25391f0fa048)

* A blue or orange light indicates the valve's comparison mode (pressure or temperature respectively)
* A light or a switch on the input pipes pointing to the side-pipe means that pipe is being read, either as control or as threshold (Couldn't really figure out a way to make it clear which).

Closes #33709

(Could have sworn there was a [content] tag, but I can't find it on the wiki)

## Why it's good
<!-- Explain why you think these changes are good. -->
Extra functionality to a new-ish toy, for supermatter and plasma cooling autism

## Changelog
<!-- See https://ss13.moe/wiki/index.php/Guide_to_Writing_a_Pull_Request -->
:cl:
 * rscadd: Pressure valves can now also compare temperature, and have been renamed to conditional valves.
 * tweak: Conditional valves can now also compare the control pipe against the input pipes' gases, rather than just a fixed value.
 * tweak: Conditional valves can now also forego the side input entirely, using either of the two other pipes as control. This might help make some designs neater and more compact.
 * rscdel: Conditional valves can no longer be controlled via AAC, beyond simple on/off commands.

Inorien commented:
Why does this need to replace the pressure valve? Though I'll admit I have never used a pressure valve so if they were really that dumb and useless then this is cool, no idea what the fuck is going on in the sprite tho that shit is complicated and huge

pedr69 commented:
it's fucking happening
non positronic AI circuits when

Eneocho commented:
Atmospherics is a subtle joke. The pipe is called "Conditional" Valves, where conditional does not both end in the sound "-ics", thus not rhyming with the name of the department, Atmospherics. The title says that the pipe was "Formerly Pressure", implying that the word beginning with "C" would've ended with "-sure". So, before this PR was opened, it would've been called "Condisure Valves".

jellyveggie2 commented:
@Inorien 
Well... At first I tried to do the RPD clean-up/QoL mentioned in #33709 but failed, and decided I would focus on the valve stuff instead. So to compensate, out of the two options discussed in that issue ("make two separate valves for temperature and pressure at the cost of RPD clutter" vs "make a single valve that can do both temperature and pressure") I choose the "single valve" option.
None of the original functionality is lost, in fact it already defaults to what used to be the original functionality (pressure mode, center control, constant threshold), so it's not that much of a replacement anyway other than the new name, since it's not just pressure anymore. 

You've got a valid point with the North/South thing and in fact that's what I first tried to do, with names updating on the UI depending on valve position, but then came updating the AAC's menu and it all started breaking apart because I would no longer have access to the valve's rotation and update the North/South/whatever inputs' direction accordingly (moreover, an instruction on the AAC can refer to multiple valves, each with their rotation).
I was going to have to come up with new terminology no matter what, and figured keeping consistent would be the lesser of to evils (doing so also simplified UI code, and I was starting to look for ways to be lazy and finish this thing already at this point). 
I will admit my naming sense sucks and I am open to alternatives.

And yeah, the new sprites do look huge now. That's what I get trying to add extra pipes in an attempt at showing where gas is being redirected. The height of my spriting skill stops at gluing pieces of other sprites together, so if an actual spriter comes in with something better I'll take it.

> Also your changelog is referring to itself (?)
...Is that really that bad? The lines are meant to be read together, they're not going to get mixed up with other changes, and I'd rather not keep repeating myself about pressure valves now being called conditional valves.

pedr69 commented:
>but then came updating the AAC's menu and it all started breaking apart

isn't the point of these things not needing an AAC script?
surely you could get a functional conditional valve(albeit much bigger) with a script and some pumps, meters and valves

Eneocho commented:
You could use these to make logic gates.
Connect the gates to the AAC, and the AAC to signalers.
Connect the signalers to light tile remotes.
You now have a screen.
Code tetris on space atmos.

jellyveggie2 commented:
>isn't the point of these things not needing an AAC script?
I made the mistake of making pressure valves compatible with the AAC thinking "why not", and I now have to live with the consequences.

Inorien commented:
>but then came updating the AAC's menu and it all started breaking apart
this sounds like suffering
yeah as others have said not sure it needs AAC, sure you could get bonkers complexity with it but may not be worth

Eneocho commented:
AAC valves please for the ultrahyperautismo.

jellyveggie2 commented:
Well, that's a 2 to 1 on dropping AAC support so far. 
I'll look into getting it done this evening unless the emojicracy votes otherwise on this comment:

Drop AAC support for conditional valves, yay or nay?

jellyveggie2 commented:
Wait I miscounted pedr69's as Eneocho changing his vote, the ratio's still 2 to 1.

Not that it changes the result I guess:
Removed AAC support for the valve (except the main enable/disable control, that one's simple enough it can stay) in favor of a better UI, and updated the PR accordingly. I'm still left with a need to refer to the pipes in a direction agnostic way, even if just for the changelog, so "side pipe" and "input pipes" will have to do for now.

Eneocho commented:
AACbros we got too cocky...

ShiftyRail commented on code/ATMOSPHERICS/components/trinary_devices/pressure_valve.dm:
```diff
@@ -76,40 +136,74 @@
+	var/input = 0
+	var/threshold = 0
+	switch(control_input)
+		if (CVALVE_INPUT_LEFT) input = get_valve_mode_gas_value(pipe_flags & IS_MIRROR ? air3 : air1)
```
Formatting, we don't do one-line `if`s, this is bad for reading; especially with the weird ternary operator after

ShiftyRail commented on code/ATMOSPHERICS/components/trinary_devices/pressure_valve.dm:
```diff
@@ -76,40 +136,74 @@
+			input = get_valve_mode_gas_value(air2)
+
+	switch(threshold_source)
+		if (CVALVE_THRESHOLD_LEFT) threshold = get_valve_mode_gas_value(pipe_flags & IS_MIRROR ? air3 : air1)
```
Ditto

ShiftyRail commented on code/ATMOSPHERICS/components/trinary_devices/pressure_valve.dm:
```diff
@@ -76,40 +136,74 @@
 
+/obj/machinery/atmospherics/trinary/pressure_valve/proc/get_valve_mode_gas_value(datum/gas_mixture/air)
+	switch(mode)
+		if (CVALVE_MODE_PRESSURE) return air.pressure
```
Ditto

ShiftyRail reviewed (changes requested):
Minor formatting changes as far as I can see

ShiftyRail commented on code/modules/atmos_automation/implementation/digital_pvalves.dm:
```diff
@@ -1,96 +1,43 @@
-		return 1
-
-// --- Pressure Valve Mode: Threshold and Above/Below ---
-/datum/automation/set_pressure_valve_mode
```
Not familiar with those, where did they go? Does it result in feature loss for the AAC?

jellyveggie2 replied:
That is right.
`/datum/automation/` are the commands that show up on the AAC. and `/datum/automation/set_pressure_valve_mode` would be the command I added when I first coded pressure valves to control their settings through the AAC. So yes, deleting that command would result in some feature loss. 
...I should note that in the changelog.

As to what led to this: #33709 brought up a request to 
>allow comparison between the control pipe (the one that is used for the conditional check) and input pipe (the one that is opened to the other side)

And that meant being able to select which input to compare the control pipe against (as there are two inputs that), which in turn meant naming each pipe. 
Applying the North/South/East/West naming scheme used in other atmos components such as gas mixers onto the valve's menu was easy enough... But it did not mesh well when it came to the AAC: as far as I can tell there's no way for me to read the controlled valve's `dir`, nor would it make sense as I must take into account the possibility that the AAC may be controlling multiple valves (each with their own `dir`).

So it's either:
* Adopt a new, `dir`-agnostic naming scheme specifically for conditional valves (my first choice), or
* Drop AAC support entirely (the choice that won out after some brief discussion in this PR)

You may want to look at [3bbba54](https://github.com/vgstation-coders/vgstation13/pull/34418/commits/3bbba54854e09b1067e4b1adca4f023538b14f0b) for a comparison between the two. I am open to alternative solutions as well.

jellyveggie2 replied:
Forgot to mention, there _is_ a third option in that I could have the AAC and the valve follow their own naming schemes each, but I really don't like the lack of consistency there.

ShiftyRail commented:
I /think/ AAC-independant valves are not too bad in and of themselves (we have manual valves for a reason), but maybe someone in the future will want to have their own way of tempering with them from a distance. We'll see.
